Develop a "Create a dapp using ethersjs" tutorial (#418)
polkadot-developers/polkadot-docs: tutorials/smart-contracts/launch-your-first-project/create-dapp-ethers-js.md

For this dApp, you'll use a simple Storage contract that's already been deployed. So you need to create an interface to interact with it. First, ensure to create a folder called `abis` in the root of your project, create a file `Storage.json`, and paste the corresponding ABI (Application Binary Interface) of the Storage contract. You can copy and paste the following:

??? code "Storage.sol ABI"

    ```json title="Storage.json"
    --8<-- "code/tutorials/smart-contracts/launch-your-first-project/create-dapp-ethers-js/Storage.json"
    ```

Now, create a file called `utils/contract.js`:

```javascript title="contract.js"
--8<-- "code/tutorials/smart-contracts/launch-your-first-project/create-dapp-ethers-js/contract.js"
```

This file defines the contract address, ABI, and functions to create instances of the contract for reading and writing.

## Create the Wallet Connection Component

Next, let's create a component to handle wallet connections. Create a new file called `components/WalletConnect.js`:

```javascript title="WalletConnect.js"
--8<-- "code/tutorials/smart-contracts/launch-your-first-project/create-dapp-ethers-js/WalletConnect.js"
```

> This component handles connecting to the wallet, switching networks if necessary, and keeping track of the connected account. To integrate this component to your dApp, you need to overwrite the existing boilerplate in `app/page.js` with the following code:

Reviewer: ```suggestion
This component handles connecting to the wallet, switching networks if necessary, and keeping track of the connected account. 

To integrate this component to your dApp, you need to overwrite the existing boilerplate in `app/page.js` with the following code:
```

In this tutorial, you'll build a complete dApp that interacts with a smart contract deployed on Asset Hub. It will use [Ethers.js](/develop/smart-contracts/libraries/ethers-js){target=\_blank} to interact with the blockchain and [Next.js](https://nextjs.org/){target=\_blank} as the frontend framework. By the end of this tutorial, you'll have a functional dApp that allows users to connect their wallets, read data from the blockchain, and execute transactions.

## Prerequisites

Before you begin, make sure you have:

- [Node.js](https://nodejs.org/en){target=\_blank} v16 or newer installed on your machine
- A crypto wallet (like MetaMask) with some test tokens. For further information, check the [Connect to Asset Hub](/develop/smart-contracts/connect-to-asset-hub){target=\_blank} guide
- Basic understanding of React and JavaScript
- Familiarity with blockchain concepts and Solidity (helpful but not mandatory)

## Project Overview

> The dApp will interact with a simple Storage contract, for more references, you can check the [Create Contracts](/tutorials/smart-contracts/launch-your-first-project/create-contracts){target=\_blank} tutorial that creates this contract step by step. This contract allows:

Reviewer 2: ```suggestion
The dApp will interact with a simple Storage contract. For a step-by-step guide on creating it, refer to the [Create Contracts](/tutorials/smart-contracts/launch-your-first-project/create-contracts){target=\_blank} tutorial. This contract allows:
```
Just to prevent word repetition (suggestion, you can modify it or enhance it)

    ├── globals.css
    ├── layout.js
    ├── page.js
    └── utils
        ├── contract.js
        └── ethers.js
```

## Set Up the Project

Let's start by creating a new Next.js project:

```
npx create-next-app ethers-dapp
cd ethers-dapp
```

Next, install the needed dependencies:

```
npm install ethers
```

## Connect to Asset Hub

To interact with Asset Hub (Westend Asset Hub in this case), you need to set up an [Ethers.js Provider](/develop/smart-contracts/libraries/ethers-js/#set-up-the-ethersjs-provider){target=\_blank} that connects to the blockchain. Create a new file called `utils/ethers.js` and add the following code:

```javascript title="ethers.js"
--8<-- "code/tutorials/smart-contracts/launch-your-first-project/create-dapp-ethers-js/ethers.js"
```
This file establishes a connection to Asset Hub and provides helper functions for obtaining a [Provider](https://docs.ethers.org/v5/api/providers/provider/){target=_blank} and [Signer](https://docs.ethers.org/v5/api/signer/){target=_blank}. The provider allows you to read data from the blockchain, while the signer enables users to send transactions and modify the blockchain state.

## Set Up the Smart Contract Interface

> For this dApp, you'll use a simple Storage contract that's already been deployed. So you need to create an interface to interact with it. First, ensure to create a folder called `abis` in the root of your project, create a file `Storage.json`, and paste the corresponding ABI (Application Binary Interface) of the Storage contract. You can copy and paste the following:

Reviewer 2: ```suggestion
For this dApp, you'll use a simple Storage contract already deployed. So, you need to create an interface to interact with it. First, ensure to create a folder called `abis` at the root of your project, create a file `Storage.json`, and paste the corresponding ABI (Application Binary Interface) of the Storage contract. You can copy and paste the following:
```

```

This component reads the `storedNumber` value from the contract and displays it to the user. It also sets up a polling interval to refresh the data periodically.

To see this change in your dApp, you need to integrate this component into the `app/page.js` file:

```javascript title="page.js"
--8<-- "code/tutorials/smart-contracts/launch-your-first-project/create-dapp-ethers-js/page.js:0:6"
--8<-- "code/tutorials/smart-contracts/launch-your-first-project/create-dapp-ethers-js/page.js:8:22"
--8<-- "code/tutorials/smart-contracts/launch-your-first-project/create-dapp-ethers-js/page.js:24:26"
```

Your dApp will automatically be updated to the following:

![](/images/tutorials/smart-contracts/launch-your-first-project/create-dapp-ethers-js/create-dapp-ethers-js-3.webp)

## Write Data to the Blockchain

Finally, let's create a component that allows users to update the stored number. Create a file called `components/WriteContract.js`:

```javascript title="WriteContract.js"
--8<-- "code/tutorials/smart-contracts/launch-your-first-project/create-dapp-ethers-js/WriteContract.js"
```

> This component allows users to input a new number and send a transaction to update the value stored in the contract. When the transaction is successful, users will see the stored value update in the ReadContract component after the transaction is confirmed.

Reviewer 2: ```suggestion
This component allows users to input a new number and send a transaction to update the value stored in the contract. When the transaction is successful, users will see the stored value update in the `ReadContract` component after the transaction is confirmed.
```
I think it's correct to add the back tick in this case, right? If not, disregard this comment.
